File: list3/java/task2/Factory.java
public final class Factory 
{   
    private Factory() throws InstantiationError
    {
        throw new InstantiationError("This is a static class you moron");
    }

    public enum ShapeSimple implements Shape_simple
    {
        CIRCLE, SQUARE, PENTAGON, HEXAGON;

        public double area(final double var1)
        {
            return create(var1).area();
        }
        public double lenght(final double var1)
        {
            return create(var1).lenght();
        }
        public String type()
        {
            return create(1).name();
        }
        private Methods create(final double var1)
        {
            switch(this)
            {
                case CIRCLE:
                    return new Circle(var1);
                case SQUARE:
                    return new Square(var1);
                case PENTAGON:
                    return new Pentagon(var1);
                case HEXAGON:
                    return new Hexagon(var1);
                default:
                    throw new IllegalArgumentException("Unable to recognize figure");
            }
        }
    }

    public enum ShapeComplex implements Shape_complex
    {
        RHOMBUS, RECTANGLE;

        public  double area(final double var1, final double var2)
        {
            return create(var1, var2).area();
        }
        public  double lenght(final double var1, final double var2)
        {
            return create(var1, var2).lenght();
        }
        public String type()
        {
            return create(1, 1).name();
        }
        private Methods create(final double var1, final double var2)
        {
            switch(this)
            {
                case RECTANGLE:
                    return new Rectangle(var1, var2);
                case RHOMBUS:
                    return new Rhombus(var1, var2);
                default:
                    throw new IllegalArgumentException("Unable to recognize figure");
            }
        }
    }
}

File: list3/cpp/source/factory.cpp
#include <factory.hpp>

#include <vector>
#include <string>
#include <stdexcept>

#include <utils.hpp>
#include <figure.hpp>
#include <circle.hpp>
#include <hexagon.hpp>
#include <pentagon.hpp>
#include <square.hpp>
#include <rectangle.hpp>
#include <rhombus.hpp>

static const Figure Factory::CreateShape(const Type type);
{
    switch (type.get_name())
    {
    case "Circle"
        return Circle(type.get_var()[0]);
    case "Hexagon"
        return Hexagon(type.get_var()[0]);
    case "Pentagon"
        return Pentagon(type.get_var()[0]);
    case "Square"
        return Square(type.get_var()[0]);
    case "Rectangle"
        return Rectangle(type.get_var()[0],type.get_var()[1]);
    case "Rhombus"
        return Rhombus(type.get_var()[0],type.get_var()[1]);
    default:
        throw std::invalid_argument(LOC()+"Unable to create shape")
    }
}